### PR TITLE
[Merged by Bors] - feat(RingTheory/Flat): definition of flat ring homomorphism and stability under composition

### PR DESCRIPTION
**Pull request by chrisflav**
Adds the definition of flat algebras, flat ring homomorphisms and shows that flatness is stable under composition.

---

The stability under composition proof can be simplified, as soon as we know that flat is equivalent to tensoring is left-exact.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Flat/Stability.lean [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+variable (R : Type u) (S : Type v) (M : Type w) (N : Type t)
+  [CommRing R] [CommRing S] [Algebra R S]
+  [AddCommGroup M] [Module R M] [Module S M] [IsScalarTower R S M]
+
+private noncomputable abbrev auxRightMul (I : Ideal R) : M ⊗[R] I →ₗ[S] M := by
+  letI i : M ⊗[R] I →ₗ[S] M ⊗[R] R := AlgebraTensorModule.map LinearMap.id I.subtype
+  letI e' : M ⊗[R] R →ₗ[S] M := AlgebraTensorModule.rid R S M
+  exact AlgebraTensorModule.rid R S M ∘ₗ i
+
+private noncomputable abbrev auxJ (I : Ideal R) : Ideal S := LinearMap.range (auxRightMul R S S I)
```
Comment (jcommelin) [inline Mathlib/RingTheory/Flat/Stability.lean:45]: Nitpick: why the `J`?

File: Mathlib/RingTheory/Flat/Stability.lean [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+open LinearMap (lsmul rTensor lTensor)
+
+open TensorProduct
+
+namespace Module.Flat
+
+variable (R : Type u) (S : Type v) (M : Type w) (N : Type t)
+  [CommRing R] [CommRing S] [Algebra R S]
+  [AddCommGroup M] [Module R M] [Module S M] [IsScalarTower R S M]
+
```
Comment (jcommelin) [inline Mathlib/RingTheory/Flat/Stability.lean:39]: Could you please write a little module docstring that explains the proof strategy?

Comment (chrisflav) [reply Mathlib/RingTheory/Flat/Stability.lean:53]: Done in the latest commit.

File: Mathlib/RingTheory/Flat/Stability.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Flat/Stability.lean:59]: Yeah, that. Or just call it `J`.

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**